# Лабораторная работа №2. Реализация глубокой нейронной сети

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import random
%matplotlib inline

## Задание 1.
Реализуйте полносвязную нейронную сеть с помощью библиотеки Tensor Flow. В качестве алгоритма оптимизации можно использовать, например, стохастический градиент (Stochastic Gradient Descent, SGD). Определите количество скрытых слоев от 1 до 5, количество нейронов в каждом из слоев до нескольких сотен, а также их функции активации (кусочно-линейная, сигмоидная, гиперболический тангенс и т.д.).

In [2]:
from utils.dataset import extract_dataset, download_dataset, read_dataset

In [3]:
dataset_url = "https://commondatastorage.googleapis.com/books1000/notMNIST_large.tar.gz"
dataset_path = extract_dataset(download_dataset(dataset_url))

In [4]:
img_h, img_w = 28, 28
classes = ["A","B","C","D","E","F","G","H","I","J"]
X, y = read_dataset(dataset_path, classes, img_h, img_w)

In [5]:
from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X.T, y.T, test_size=0.25, random_state=42)

In [15]:
import tensorflow as tf

# model hyper parameters
num_epochs = 50
batch_size = 1024

model = tf.keras.models.Sequential([
    tf.keras.layers.Dense(128, activation='relu'),
    tf.keras.layers.Dense(128, activation='relu'),
    tf.keras.layers.Dense(10, activation='softmax')
])

optimizer = tf.keras.optimizers.SGD(lr=0.01)

model.compile(optimizer=optimizer,
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])
    
model.fit(X_train, Y_train, epochs=num_epochs, batch_size=batch_size)
model.evaluate(X_test, Y_test)

Train on 396835 samples
Epoch 1/50
396835/396835 [==============================] - 2s 6us/sample - loss: 1.1973 - accuracy: 0.6922
Epoch 2/50
396835/396835 [==============================] - 2s 6us/sample - loss: 0.7276 - accuracy: 0.8046
Epoch 3/50
396835/396835 [==============================] - 2s 6us/sample - loss: 0.6683 - accuracy: 0.8172
Epoch 4/50
396835/396835 [==============================] - 3s 6us/sample - loss: 0.6386 - accuracy: 0.8240
Epoch 5/50
396835/396835 [==============================] - 3s 7us/sample - loss: 0.6187 - accuracy: 0.8282
Epoch 6/50
396835/396835 [==============================] - 3s 6us/sample - loss: 0.6033 - accuracy: 0.8317
Epoch 7/50
396835/396835 [==============================] - 3s 6us/sample - loss: 0.5905 - accuracy: 0.8343
Epoch 8/50
396835/396835 [==============================] - 3s 7us/sample - loss: 0.5796 - accuracy: 0.8364
Epoch 9/50
396835/396835 [==============================] - 3s 8us/sample - loss: 0.5699 - accuracy: 0.8385
Epoc

[0.4368418463719057, 0.8703649]

## Задание 2.
Как улучшилась точность классификатора по сравнению с логистической регрессией?


Точность увеличилась, с помощью логистической регрессии мы добились результата 81%. С помощью нейронной сети точность составила: 87%.

## Задание 3.
Используйте регуляризацию и метод сброса нейронов (dropout) для борьбы с переобучением. Как улучшилось качество классификации?


In [13]:
# model hyper parameters
dropout_rate = 0.1
l2_regularization = 1e-4

model=tf.keras.models.Sequential([
    tf.keras.layers.Dense(128, activation='relu',kernel_regularizer=tf.keras.regularizers.l2(l2_regularization)),
    tf.keras.layers.Dropout(dropout_rate),
    tf.keras.layers.Dense(128, activation='relu',kernel_regularizer=tf.keras.regularizers.l2(l2_regularization)),
    tf.keras.layers.Dropout(dropout_rate),
    tf.keras.layers.Dense(10, activation='softmax')
])

optimizer = tf.keras.optimizers.SGD(lr=0.01)

model.compile(optimizer=optimizer,
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

model.fit(X_train, Y_train, epochs=num_epochs, batch_size=batch_size)
model.evaluate(X_test, Y_test)

Train on 396835 samples
Epoch 1/50
396835/396835 [==============================] - 3s 9us/sample - loss: 1.3540 - accuracy: 0.6187
Epoch 2/50
396835/396835 [==============================] - 3s 7us/sample - loss: 0.8384 - accuracy: 0.7792
Epoch 3/50
396835/396835 [==============================] - 3s 7us/sample - loss: 0.7543 - accuracy: 0.7999
Epoch 4/50
396835/396835 [==============================] - 3s 7us/sample - loss: 0.7118 - accuracy: 0.8101
Epoch 5/50
396835/396835 [==============================] - 3s 7us/sample - loss: 0.6851 - accuracy: 0.8153
Epoch 6/50
396835/396835 [==============================] - 3s 7us/sample - loss: 0.6652 - accuracy: 0.8204
Epoch 7/50
396835/396835 [==============================] - 3s 7us/sample - loss: 0.6500 - accuracy: 0.8227
Epoch 8/50
396835/396835 [==============================] - 3s 7us/sample - loss: 0.6371 - accuracy: 0.8262
Epoch 9/50
396835/396835 [==============================] - 3s 8us/sample - loss: 0.6274 - accuracy: 0.8285
Epoc

[0.4732068698067496, 0.86924607]

Как видим, качество классификации осталось примерно таким же. Скорее всего, наша сеть не переобучилась и в прошлый раз.

## Задание 4.
Воспользуйтесь динамически изменяемой скоростью обучения (learning rate). Наилучшая точность, достигнутая с помощью данной модели составляет 97.1%. Какую точность демонстрирует Ваша реализованная модель?


In [16]:
# model hyper parameters
dropout_rate = 0.1
l2_regularization = 1e-4

model=tf.keras.models.Sequential([
    tf.keras.layers.Dense(128, activation='relu',kernel_regularizer=tf.keras.regularizers.l2(l2_regularization)),
    tf.keras.layers.Dropout(dropout_rate),
    tf.keras.layers.Dense(128, activation='relu',kernel_regularizer=tf.keras.regularizers.l2(l2_regularization)),
    tf.keras.layers.Dropout(dropout_rate),
    tf.keras.layers.Dense(10, activation='softmax')
])

model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

model.fit(X_train, Y_train, epochs=num_epochs, batch_size=batch_size)
model.evaluate(X_test, Y_test)

Train on 396835 samples
Epoch 1/50
396835/396835 [==============================] - 3s 9us/sample - loss: 0.6351 - accuracy: 0.8246
Epoch 2/50
396835/396835 [==============================] - 3s 7us/sample - loss: 0.4707 - accuracy: 0.8700
Epoch 3/50
396835/396835 [==============================] - 3s 7us/sample - loss: 0.4297 - accuracy: 0.8815
Epoch 4/50
396835/396835 [==============================] - 3s 7us/sample - loss: 0.4079 - accuracy: 0.8872
Epoch 5/50
396835/396835 [==============================] - 3s 7us/sample - loss: 0.3929 - accuracy: 0.8918
Epoch 6/50
396835/396835 [==============================] - 3s 7us/sample - loss: 0.3817 - accuracy: 0.8949
Epoch 7/50
396835/396835 [==============================] - 3s 8us/sample - loss: 0.3734 - accuracy: 0.8971
Epoch 8/50
396835/396835 [==============================] - 3s 7us/sample - loss: 0.3670 - accuracy: 0.8989
Epoch 9/50
396835/396835 [==============================] - 3s 8us/sample - loss: 0.3594 - accuracy: 0.9011
Epoc

[0.3306251625201281, 0.9120873]

Точность 91% при использовании Adam оптимизатора. Это показывает то, что возможно мы не лучшим образом подобрали параметры для схохастического градиентного спуска.